v3/MmlTree/ts/Visitor.ts

```diff
@@ -0,0 +1,173 @@
+     */
+    visitTree(tree: Node, ...args: any[]): any;
+
+    /*
+     * Visit a node by calling the visitor function for the givn type of node
+     *  (passing along any needed parameters)
+     *
+     * @param {Node} node   The node to visit
+     * @param {any[]} args  The arguments to pass to the visitor function for this node
+     * @return {any}        Whatever the visitor function returns for this node
+     */
+    visitNode(node: Node, ...args: any[]): any;
+
+    /*
+     * The default visitor function for when no node-specific function is defined
+     *
+     * @param {Node} node   The node to visit
+     * @param {any[]} args  The arguments to pass to the visitor function for this node
+     * @return {any}        Whatever the visitor function returns for this node
+     */
+    visitDefault(node: Node, ...args: any[]): any;
+
+    /*
+     * Define a visitor function for a given node kind
+     *
+     * @param {string} kind  The node kind for which the handler is being defined
+     * @param {VisitorFunction} handler  The function to call to handle nodes of this kind
+     */
+    setNodeHandler(kind: string, handler: VisitorFunction): void;
+
+    /*
+     * Remove the visitor function for a given node kind
+     *
+     * @param {string} kind  The node kind whose visitor function is to be removed
+     */
+    removeNodeHandler(kind: string): void;
+
+    /*
+     * The various visitor functions implemented by the subclasses, and any data they need
+     */
+    [property: string]: any;
+}
+
+/*****************************************************************/
+/*
+ *  Implements the generic Visitor object
+ */
+
+export class Visitor implements IVisitor {
+    /*
+     * Holds the mapping from node kinds to visitor funcitons
+     */
+    protected nodeHandlers: Map<string, VisitorFunction> = new Map();
+
+    /*
+     *  Visitor functions are named "visitKindNode" where "Kind" is replaced by
+     *    the node kind; e.g., visitTextNode for kind = text.
+     *
+     *  @param {string} kind  The node kind whose method name is needed
+     *  @return {string}  The name of the visitor method for the given node kind
+     */
+    protected static methodName(kind: string) {
+        return 'visit' + kind.charAt(0).toUpperCase() + kind.substr(1) + 'Node';
+    }
+
+    /*
+     * Create the node handler map by looking for methods with the correct names
+     *   based on the node kinds available from the factory.
+     *
+     * @constructor
+     * @param {NodeFactory} factory  The node factory for the kinds of nodes this visitor handles
+     */
+    constructor(factory: NodeFactory) {
+        for (const kind of factory.getKinds()) {
+            let method = (this as IVisitor)[Visitor.methodName(kind)] as VisitorFunction;
+            if (method) {
+                this.nodeHandlers.set(kind, method);
+            }
+        }
+    }
+
+    /*
+     * Visit the tree by visiting the root node
```

Comment:
I don't really know what `@override` does, or what is needed for it.  I would think that in some cases, a new comment would be needed to explain what the overridden function does (i.e., what its additional functionality is that isn't in the original).